packages/@react-stately/numberfield/src/useNumberFieldState.ts

```diff
@@ -11,133 +11,298 @@
+  let numberParser = useNumberParser(inputValueFormatter, numeralOverride);
+  let intlOptions = useMemo(() => inputValueFormatter.resolvedOptions(), [inputValueFormatter]);
+
+  let isMaxRange = minValue === Number.MIN_SAFE_INTEGER && maxValue === Number.MAX_SAFE_INTEGER;
+
+  let {minusSign, plusSign, decimal, validCharacters, literals, group} = numberParser.symbols;
+  minusSign = minValue >= 0 || !minusSign ? '' : minusSign;
+  plusSign = maxValue <= 0 || !plusSign ? '' : plusSign;
+
+  // javascript doesn't recognize NaN === NaN, so multiple onChanges will get fired if we don't ignore consecutive ones
+  // in addition, if the input starts with a number, then we'll count that as the last val dispatched, we only need to calculate it the first time
+  let startingValue = useMemo(() => {
+    if (!isNaN(value)) {
+      return value;
+    } else if (!isNaN(defaultValue)) {
+      return defaultValue;
+    }
+    return NaN;
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, []);
+  let lastValDispatched = useRef(startingValue);
+  let smartOnChange = useCallback((val) => {
+    if (!isNaN(val) || !isNaN(lastValDispatched.current)) {
+      onChange?.(val);
+    }
+    lastValDispatched.current = val;
+  }, [lastValDispatched, onChange]);
 
-  const numberParser = useNumberParser();
-  const textValueFormatter = useNumberFormatter(formatOptions);
-  const inputValueFormatter = useNumberFormatter();
 
-  const [numberValue, setNumberValue] = useControlledState<number>(value, defaultValue || 0, onChange);
-  let initialInputValue = inputValueFormatter.format(numberValue);
-  const [inputValue, setInputValue] = useState(isNaN(value) && isNaN(defaultValue) ? '' : initialInputValue);
-  const [isValid, setIsValid] = useState(isInputValueValid(numberValue, maxValue, minValue));
+  let [numberValue, setNumberValue] = useControlledState<number>(value, isNaN(defaultValue) ? NaN : defaultValue, smartOnChange);
 
-  const minusSign = useRef('-');
+  let initialInputValue = isNaN(numberValue) ? '' : inputValueFormatter.format(numberValue);
+  let [inputValue, setInputValue] = useState(initialInputValue);
 
+  let textValue = inputValueFormatter.format(numberValue);
+
+  // this updates the field only if the formatter or number has changed,
+  // this should only run after increment, decrement, blur
+  // or when the formatter changes, but not because of the overrides, which is why we run on formatOptions
+  // I doubt that the formatOptions will change while a user is typing a number, but that might cause a problem
   useEffect(() => {
-    // Get the minus sign of the current locale to filter the input value
-    // Automatically updates the minus sign when numberFormatter changes
-    minusSign.current = inputValueFormatter.formatToParts(-11).find(p => p.type === 'minusSign').value;
-  }, [inputValueFormatter]);
+    setInputValue(isNaN(numberValue) ? '' : textValue);
```

Comment:
IMO it would probably be better to keep a ref with the last value of the values you care about and compare them in an if statement rather than relying on the hook dependency array. I believe React can call hooks in other scenarios than just when the dep array changes, e.g. on fast refresh or in concurrent mode, so we can't rely on that alone.

Reply:
fascinating, didn't realize that, thanks